javaslang/src/main/java/javaslang/collection/Tree.java

```diff
@@ -748,12 +771,12 @@ public boolean isEmpty() {
 
         @Override
         public boolean isLeaf() {
-            return children.isEmpty();
+            return size == 1;
         }
 
         @Override
         public int size() {
-            return 1 + children.foldLeft(0, (acc, child) -> acc + child.length());
+            return size;
```

Comment:
@ruslansennov 

> It is size of object. I believe that reference's size is smaller :)

sounds good

@pivovarit 

Many thanks for your investigation! for now we should not optimize Lazy - the current impl is bulletproof and easy. @jbgi also made some basic benchmarks comparing different methods for lazy evaluation (also used in scalaz) and it turned out that the current impl is as fast as the other methods.

I'm not happy to increase the GC significantly. GC implicitly decreases the overall time-performance of the application. So the time-benefit of using Lazy instances may be reduced by having more time-consuming GCs.

I think it depends on the use-case what is the best solution. Also: How probable is it that size() is called? If the internal implementation of the collection methods will depend on size() calls (we've done that for optimizations) then it comes handy to have size already computed (we save memory and GC). Also when size() is not called at all in some applications then the Lazy instances are unnecessary.

Having a Lazy size is an optimization. Maybe it is premature because we currently have no reason to optimize the implementation. I like to implement the simplest solution possible and optimize when necessary.